packages/eip-7702-internal-rpc-middleware/src/utils.ts

```diff
@@ -0,0 +1,85 @@
+import { providerErrors, rpcErrors } from '@metamask/rpc-errors';
+import type { Struct, StructError } from '@metamask/superstruct';
+import { validate } from '@metamask/superstruct';
+import type { Hex } from '@metamask/utils';
+import { isHexAddress } from '@metamask/utils';
+
+/**
+ * Validates address format, checks user eth_accounts permissions, and normalizes to lowercase.
+ *
+ * @param address - The Ethereum address to validate and normalize.
+ * @param origin - The origin string for permission checking.
+ * @param getPermittedAccountsForOrigin - Function to retrieve permitted accounts for the origin.
+ * @returns A normalized (lowercase) hex address if valid and authorized.
+ * @throws JsonRpcError with unauthorized error if the requester doesn't have permission to access the address.
+ * @throws JsonRpcError with invalid params if the address format is invalid.
+ */
+export async function validateAndNormalizeAddress(
+  address: Hex,
+  origin: string,
+  getPermittedAccountsForOrigin: (origin: string) => Promise<string[]>,
+): Promise<Hex> {
+  if (
+    typeof address === 'string' &&
+    address.length > 0 &&
+    isHexAddress(address)
+  ) {
+    // Ensure that an "unauthorized" error is thrown if the requester
+    // does not have the `eth_accounts` permission.
+    const accounts = await getPermittedAccountsForOrigin(origin);
+
+    // Validate and convert each account address to normalized Hex
+    const normalizedAccounts: string[] = accounts.map((accountAddress) =>
+      accountAddress.toLowerCase(),
+    );
+
+    const normalizedAddress = address.toLowerCase();
+
+    if (normalizedAccounts.includes(normalizedAddress)) {
+      // we know that normalizedAddress is a valid Hex string because of isHexAddress
+      return normalizedAddress as Hex;
+    }
+
+    throw providerErrors.unauthorized();
+  }
+
+  throw rpcErrors.invalidParams({
+    message: `Invalid parameters: must provide an Ethereum address.`,
+  });
```

Comment:
nit, is it okay if we inverse the if statement at the start of this function and then throw for an early exit?

Reply:
`must provide a valid Hex address` Maybe?